### DISCUSSION TITLE
Could this library be used with Hive Metastore Standalone with Kerberos support

### DISCUSSION THREAD

**Discussion by richardwooding** [beltran/gohive#139] [Q&A] [answered]
Hive Metastore can be installed in a standalone fashion. Could this this library be used to connect to Hive Metastore Standalone, or does it expect the full Hive ecosystem. Additionally, in our configuration our server principal is a different value from our client principal, is it is possible to configure Kerberos authentication in this fashion.

i.e. our value for hive.metastore.kerberos.principal is different from the principal used with kinit.

We have this working in a JVM with **org.apache.hive:hive-standalone-metastore:3.1.2** library, and we where wondering if we could do the same but in go.

**Comment by beltran** [comment, marked as answer]
Hello, unfortunately there's not a client for the hive metastore in this package. I'm not aware of any hive metastore libraries in go that also support kerberos but I guess there could be one out there. This is something I've wanted to do but haven't had the time yet. I may add it in the next weeks/months.

> i.e. our value for hive.metastore.kerberos.principal is different from the principal used with kinit.

Then are you setting some values in the metastore property `hadoop.proxyuser.hive.groups` that allows the kinit user to impersonate the metastore user? If this is the case the metastore client shouldn't have to worry about the kinit user and the metastore user being different.

**Reply by richardwooding** [reply]
Thank you for your swift response. I confirmed that we do indeed have specific settings in `hadoop.proxyuser.hive.groups` as well.

[a comment by an automated account is not shown]

**Comment by beltran** [comment]
I did try to get kerberos to work with the metastore client. See [here](https://github.com/beltran/gohive/blob/92aef12e4b46fd9d2c3c63219594477f9dc92852/gohivemeta/gohivemeta.go). The whole branch is https://github.com/beltran/gohive/tree/metastore_client. It was working with non-kerberos auth. I couldn't get it work with kerberos.

>  I think it is not difficult to add kerberos support to gometastore, because hive metastore RPC communication with kerberos should be similar to gohive do, am I right?

This was my expectation as well. This was years ago, but I remember I was getting an error and simply didn't know how to proceed.

**Comment by beltran** [comment]
hmm, I actually think it's working now https://github.com/beltran/gohive/tree/metastore_client. I added an example of how to use the metastore client with kerberos, it can be run with `go run -tags "kerberos integration" example/main_meta.go`. If you do run it please let me know if it works or you run into any issues.

I'll merge the PR in the next days once I clean it up.

**Reply by beltran** [reply]
Happy to help, I've merged the PR, please see the docs: https://github.com/beltran/gohive?tab=readme-ov-file#connection-to-the-hive-metastore

**Reply by beltran** [reply]
> The following things are easy, let me complete the left functions of HiveMetastoreClient.

That would be great!

> why we dont need to set Service Principle Name of kerberos? 

Not sure I follow, where would you set it? As long as `kinit` has been done, gohive will pick up everything it needs.

> And is there any document( I did search by google but not found) about communication details about client connect to hiveserver2/hive metastore thrift + kerberos?

I don't think there are great docs. When I implemented this I went over the java library implementation and https://datatracker.ietf.org/doc/html/rfc4752 (GSSAPI), but I haven't checked for docs in a while.

[a comment by an automated account is not shown]

**Comment by beltran** [comment]
Closing the discussion since querying metastore is supported now.
